src/main/java/org/jabref/gui/mergeentries/MergeEntriesHelper.java

```diff
@@ -0,0 +1,128 @@
+    private MergeEntriesHelper() {
+        // Private constructor to prevent instantiation of utility class
+    }
+
+    /**
+     * Merges two BibEntry objects with undo support.
+     * Use this method when modifying entries directly in the UI.
+     * The original entry will be updated with information from the fetched entry.
+     */
+    public static void mergeEntries(BibEntry originalEntry, BibEntry fetchedEntry, NamedCompound undoManager) {
+        Objects.requireNonNull(originalEntry, "Original entry cannot be null");
+        Objects.requireNonNull(fetchedEntry, "Fetched entry cannot be null");
+        Objects.requireNonNull(undoManager, "Undo manager cannot be null");
+
+        LOGGER.debug("Starting merge of entries. Original type: {}, Fetched type: {}",
+                originalEntry.getType(), fetchedEntry.getType());
+
+        updateEntryTypeIfDifferent(originalEntry, fetchedEntry, undoManager);
+        updateFieldsWithNewInfo(originalEntry, fetchedEntry, undoManager);
+        removeObsoleteFields(originalEntry, fetchedEntry, undoManager);
+
+        LOGGER.debug("Finished merging entries");
+    }
+
+    private static void updateEntryTypeIfDifferent(final BibEntry originalEntry,
+                                                   final BibEntry fetchedEntry, final NamedCompound undoManager) {
+        EntryType oldType = originalEntry.getType();
+        EntryType newType = fetchedEntry.getType();
+
+        if (!oldType.equals(newType)) {
+            LOGGER.debug("Updating entry type from {} to {}", oldType, newType);
+            originalEntry.setType(newType);
+            undoManager.addEdit(new UndoableChangeType(originalEntry, oldType, newType));
+        }
+    }
+
+    private static void updateFieldsWithNewInfo(final BibEntry originalEntry,
+                                                final BibEntry fetchedEntry, final NamedCompound undoManager) {
+        Set<Field> jointFields = getJointFields(originalEntry, fetchedEntry);
+        LOGGER.debug("Processing {} joint fields for updates", jointFields.size());
+
+        for (Field field : jointFields) {
+            updateFieldIfNecessaryWithUndo(originalEntry, fetchedEntry, field, undoManager);
+        }
+    }
+
+    private static void updateFieldIfNecessaryWithUndo(final BibEntry originalEntry,
+                                                       final BibEntry fetchedEntry,
+                                                       final Field field, final NamedCompound undoManager) {
+        fetchedEntry.getField(field).ifPresent(fetchedValue -> originalEntry.getField(field).ifPresentOrElse(
+                originalValue -> {
+                    if (fetchedValue.length() > originalValue.length()) {
+                        updateField(originalEntry, field, originalValue, fetchedValue, undoManager);
+                    }
+                },
+                () -> updateField(originalEntry, field, null, fetchedValue, undoManager)
+        ));
+    }
```

Comment:
This method could be simplified for better readability. I'd suggest using `.map` and `.flatMap` as suggested by the maintainers.

Reply:
Sure, will optimize soon